File: Objective-C Sample Code/Chapter 10/Going Further/Top10/Top10/MasterViewController.m
//
//  MasterViewController.m
//  Top10
//
//  Created by Cory Bohon on 12/5/16.
//  Copyright © 2016 Cory Bohon. All rights reserved.
//

#import "MasterViewController.h"
#import "DetailViewController.h"
#import "JSONItem.h"
#import "NetworkController.h"

@interface MasterViewController ()

@property (nonatomic, strong) NSArray *objects;

@end

@implementation MasterViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.title = @"Top 10 Tunes";
    [self reloadContent: self];
}

#pragma mark - Segues 
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([[segue identifier] isEqualToString:@"showDetail"]) {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        JSONItem *detailItem = [self.objects objectAtIndex:indexPath.row];
        
        DetailViewController *detailController = (DetailViewController *)[[[segue destinationViewController] viewControllers] firstObject];
        
        [detailController setDetailItem:detailItem];
    }
}

#pragma mark - Actions 
- (void)reloadContent:(id)sender
{
    [[NetworkController sharedNetworkController] retrieveJSONFeedWithCompletionHandler:^(NSError *error, NSArray *objects) {
        if (!error && objects) {
            self.objects = objects;
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.tableView reloadData];
            });
        }
    } forNumberOfItems:10];
}

#pragma mark - Table View
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.objects count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    JSONItem *item = [self.objects objectAtIndex:indexPath.row];
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    cell.textLabel.text = [NSString stringWithFormat:@"%ld - %@", (long)indexPath.row + 1, item.songTitle];
    
    return cell;
}

@end

File: Swift Sample Code/Chapter 9/Going Further/Top10/Top10/MasterViewController.swift
//
//  MasterViewController.swift
//  Top10
//
//  Created by Cory Bohon on 2/13/17.
//  Copyright © 2017 Cory Bohon. All rights reserved.
//

import UIKit

class MasterViewController: UITableViewController {
    
    var objects: Array<Any>?

    override func viewDidLoad() {
        self.title = "Top 10 Tunes"
        self.reloadContent(sender: self)
    }
    
    // MARK: - Segues
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "showDetail" {
            
            guard self.tableView.indexPathForSelectedRow != nil else {
                return
            }
            
            let indexPath: IndexPath = self.tableView.indexPathForSelectedRow!
            let detailItem: JSONItem = self.objects![indexPath.row] as! JSONItem
            
            let navigationController = segue.destination as! UINavigationController
            let destinationController = navigationController.viewControllers.first as! DetailViewController
            
            let detailController: DetailViewController = destinationController
            detailController.detailItem = detailItem
        }
    }
    
    @IBAction func reloadContent(sender: AnyObject) {
        NetworkController.shared().retrieveJSONFeed(completionHandler: { (error, objects) in
            
            self.objects = objects
            
            DispatchQueue.main.async {
                self.tableView.reloadData()
            }
        }, forNumberOfItems: 10)
    }
    
    // MARK: - Table View
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        guard self.objects != nil else {
            return 0
        }
        
        return self.objects!.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let item: JSONItem = self.objects![indexPath.row] as! JSONItem
        
        let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "Cell")!
        cell.textLabel?.text = "\(indexPath.row + 1) - \(item.songTitle!)"
        
        return cell
    }
    
    
}

